File: test/room_activity_session_contract_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nudge/models/room_activity_session.dart';

void main() {
  final startedAt = DateTime.utc(2026, 7, 27, 9);

  test('member starts and owns their own room activity session', () {
    final session = RoomActivitySession.start(
      sessionId: 'session-alice-focus',
      roomId: 'room-study',
      actorId: 'alice',
      activityKind: RoomActivityKind.focus,
      metricUnit: 'minutes',
      targetValue: 25,
      source: RoomActivitySource.app,
      now: startedAt,
    );

    expect(session.status, RoomActivitySessionStatus.active);
    expect(session.metricValue, 0);
    expect(session.actorId, 'alice');
    expect(session.toJson()['activityKind'], 'focus');
  });

  test('member can pause, resume, and complete without a room owner', () {
    final started = RoomActivitySession.start(
      sessionId: 'session-alice-exercise',
      roomId: 'room-exercise',
      actorId: 'alice',
      activityKind: RoomActivityKind.exercise,
      metricUnit: 'minutes',
      targetValue: 30,
      source: RoomActivitySource.web,
      now: startedAt,
    );

    final paused = started.transition(
      actorId: 'alice',
      nextStatus: RoomActivitySessionStatus.paused,
      metricValue: 12,
      now: startedAt.add(const Duration(minutes: 12)),
    );
    final resumed = paused.transition(
      actorId: 'alice',
      nextStatus: RoomActivitySessionStatus.active,
      metricValue: 12,
      now: startedAt.add(const Duration(minutes: 15)),
    );
    final completed = resumed.transition(
      actorId: 'alice',
      nextStatus: RoomActivitySessionStatus.completed,
      metricValue: 30,
      now: startedAt.add(const Duration(minutes: 33)),
    );

    expect(paused.status, RoomActivitySessionStatus.paused);
    expect(resumed.status, RoomActivitySessionStatus.active);
    expect(completed.status, RoomActivitySessionStatus.completed);
    expect(completed.endedAt, isNotNull);
  });

  test('room owner cannot transition another member session', () {
    final session = RoomActivitySession.start(
      sessionId: 'session-alice-steps',
      roomId: 'room-steps',
      actorId: 'alice',
      activityKind: RoomActivityKind.steps,
      metricUnit: 'steps',
      targetValue: 8000,
      source: RoomActivitySource.health,
      now: startedAt,
    );

    expect(
      () => session.transition(
        actorId: 'room-owner',
        nextStatus: RoomActivitySessionStatus.completed,
        metricValue: 8000,
        now: startedAt.add(const Duration(hours: 8)),
      ),
      throwsStateError,
    );
  });

  test('terminal sessions cannot resume and progress cannot decrease', () {
    final started = RoomActivitySession.start(
      sessionId: 'session-alice-sleep',
      roomId: 'room-sleep',
      actorId: 'alice',
      activityKind: RoomActivityKind.sleep,
      metricUnit: 'hours',
      targetValue: 7,
      source: RoomActivitySource.health,
      now: startedAt,
    );
    final completed = started.transition(
      actorId: 'alice',
      nextStatus: RoomActivitySessionStatus.completed,
      metricValue: 7.5,
      now: startedAt.add(const Duration(hours: 8)),
    );

    expect(
      () => completed.transition(
        actorId: 'alice',
        nextStatus: RoomActivitySessionStatus.active,
        metricValue: 7.5,
        now: startedAt.add(const Duration(hours: 9)),
      ),
      throwsStateError,
    );
    expect(
      () => started.transition(
        actorId: 'alice',
        nextStatus: RoomActivitySessionStatus.paused,
        metricValue: -1,
        now: startedAt.add(const Duration(minutes: 1)),
      ),
      throwsArgumentError,
    );
  });
}

File: test/room_activity_session_contract_test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const contract = require("../web_dashboard/assets/room_activity_session_contract.js");

const startedAt = "2026-07-27T09:00:00.000Z";

test("web room session contract follows member-controlled transitions", () => {
  const started = contract.start({
    sessionId: "session-alice-focus",
    roomId: "room-study",
    actorId: "alice",
    activityKind: "focus",
    metricUnit: "minutes",
    targetValue: 25,
    source: "web",
    now: startedAt,
  });
  const paused = contract.transition(started, {
    actorId: "alice",
    nextStatus: "paused",
    metricValue: 10,
    now: "2026-07-27T09:10:00.000Z",
  });
  const completed = contract.transition(paused, {
    actorId: "alice",
    nextStatus: "completed",
    metricValue: 25,
    now: "2026-07-27T09:28:00.000Z",
  });

  assert.equal(started.status, "active");
  assert.equal(paused.status, "paused");
  assert.equal(completed.status, "completed");
  assert.equal(completed.endedAt, "2026-07-27T09:28:00.000Z");
});

test("web contract rejects owner control and terminal replay", () => {
  const started = contract.start({
    sessionId: "session-alice-steps",
    roomId: "room-steps",
    actorId: "alice",
    activityKind: "steps",
    metricUnit: "steps",
    targetValue: 8000,
    source: "health",
    now: startedAt,
  });
  assert.throws(() =>
    contract.transition(started, {
      actorId: "room-owner",
      nextStatus: "completed",
      metricValue: 8000,
      now: "2026-07-27T18:00:00.000Z",
    }),
  );

  const completed = contract.transition(started, {
    actorId: "alice",
    nextStatus: "completed",
    metricValue: 8000,
    now: "2026-07-27T18:00:00.000Z",
  });
  assert.throws(() =>
    contract.transition(completed, {
      actorId: "alice",
      nextStatus: "active",
      metricValue: 8000,
      now: "2026-07-27T18:01:00.000Z",
    }),
  );
});

test("protected health rooms never accept manual Web activity control", () => {
  assert.equal(
    contract.requiresTrustedHealthAdapter({ goalSourceType: "steps" }),
    true,
  );
  assert.equal(
    contract.requiresTrustedHealthAdapter({ goalSourceType: "sleepHours" }),
    true,
  );
  assert.equal(
    contract.requiresTrustedHealthAdapter({
      goalSourceType: "exerciseMinutes",
    }),
    true,
  );
  assert.equal(
    contract.requiresTrustedHealthAdapter({ goalSourceType: "studyRoom" }),
    false,
  );
});

test("web restores room UI state from the canonical Activity Ledger session", () => {
  const session = contract.fromCanonicalLedger(
    {
      activitySessionId: "session-alice-focus",
      actorUserId: "alice",
      activityType: "focus",
      status: "paused",
      source: "web",
      roomIds: ["room-study"],
      roomTargetValue: 50,
      startedAt: "2026-07-27T09:00:00.000Z",
      updatedAt: "2026-07-27T09:25:00.000Z",
      endedAt: null,
      metricValue: 25,
      metricUnit: "minutes",
      sourceSessionIds: ["session-alice-focus"],
    },
    "room-study",
  );

  assert.deepEqual(session, {
    schemaVersion: 1,
    sessionId: "session-alice-focus",
    roomId: "room-study",
    actorId: "alice",
    activityKind: "focus",
    metricUnit: "minutes",
    targetValue: 50,
    metricValue: 25,
    source: "web",
    status: "paused",
    startedAt: "2026-07-27T09:00:00.000Z",
    updatedAt: "2026-07-27T09:25:00.000Z",
    endedAt: null,
  });
  assert.throws(() =>
    contract.fromCanonicalLedger(
      {
        ...session,
        activitySessionId: session.sessionId,
        actorUserId: session.actorId,
        activityType: session.activityKind,
        roomIds: [session.roomId],
        roomTargetValue: session.targetValue,
      },
      "room-other",
    ),
  );
});
